components/mdns/mdns.c

```diff
@@ -6797,4 +6926,604 @@ void mdns_debug_packet(const uint8_t *data, size_t len)
+                !_str_null_or_empty(r->instance_name) && !strcasecmp(instance, r->instance_name) &&
+                !_str_null_or_empty(r->service_type) && !strcasecmp(service, r->service_type) &&
+                !_str_null_or_empty(r->proto) && !strcasecmp(proto, r->proto)) {
+            if (r->txt) {
+                // If the result has a previous txt entry, we delete it and re-add.
+                // TODO: we need to check if txt changes.
+                for (size_t i = 0; i < r->txt_count; i++) {
+                    free((char *)(r->txt[i].key));
+                    free((char *)(r->txt[i].value));
+                }
+                free(r->txt);
+                free(r->txt_value_len);
+            }
+            r->txt = txt;
+            r->txt_value_len = txt_value_len;
+            r->txt_count = txt_count;
+            if (r->ttl != ttl) {
+                uint32_t previous_ttl = r->ttl;
+                if (r->ttl == 0) {
+                    r->ttl = ttl;
+                } else {
+                    _mdns_result_update_ttl(r, ttl);
+                }
+                if (previous_ttl != r->ttl) {
+                    if (_mdns_browse_sync_result_find_or_add_result(out_sync_browse, r) != ESP_OK) {
+                        return;
+                    }
+                }
+            }
+            return;
+        }
+        r = r->next;
+    }
+    r = (mdns_result_t *)malloc(sizeof(mdns_result_t));
+    if (!r) {
+        HOOK_MALLOC_FAILED;
+        goto free_txt;
+    }
+    memset(r, 0, sizeof(mdns_result_t));
+    r->instance_name = strdup(instance);
+    r->service_type = strdup(service);
+    r->proto = strdup(proto);
+    if (!r->instance_name || !r->service_type || !r->proto) {
+        free(r->instance_name);
+        free(r->service_type);
+        free(r->proto);
+        free(r);
+        return;
+    }
+    r->txt = txt;
+    r->txt_value_len = txt_value_len;
+    r->txt_count = txt_count;
+    r->esp_netif = _mdns_get_esp_netif(tcpip_if);
+    r->ip_protocol = ip_protocol;
+    r->ttl = ttl;
+    r->next = browse->result;
+    browse->result = r;
+    _mdns_browse_sync_result_find_or_add_result(out_sync_browse, r);
+    return;
+
+free_txt:
+    for (size_t i = 0; i < txt_count; i++) {
+        free((char *)(txt[i].key));
+        free((char *)(txt[i].value));
+    }
+    free(txt);
+    free(r->txt_value_len);
+    return;
+}
+
+static esp_err_t _mdns_copy_address_in_previous_result(mdns_result_t *result_list, mdns_result_t *r)
+{
+    mdns_ip_addr_t *r_addr;
+    mdns_ip_addr_t *prv = NULL;
+    while (result_list) {
+        if (!_str_null_or_empty(result_list->hostname) && !_str_null_or_empty(r->hostname) && !strcasecmp(result_list->hostname, r->hostname) &&
+                result_list->ip_protocol == r->ip_protocol && result_list->addr) {
+            // If there is a same hostname in previous hostname, we need to copy the address here.
```

Comment:
Question: Why we'd find a hostname in the current result with no address?

Reply:
Because one host may pubilsh more than one `_service._protocal` with different instance names, and in the result, we might record them in separated place.

For example, host will publish these mdns service:
```
hostname.test._test._udp
hostname.test1._test._udp
hostname.test2._test._udp
```
When we need to add the `hostname.test1._test._udp` to the result, we need copy the address from the result `hostname.test._test._udp`.

Reply:
Oh, I see, it's actually not a hostname, but kind of a host-namespace of a service.
Thanks for explaining!